# Data 200: Data Systems for Data Analytics (Spring 2025)

# Homework 4b: Advanced Operations in Pandas


<font color='red'>**Due Date and Time:** 11:59 pm on Monday, 2/17/2025 </font>
---
Enter your name in the markdown cell below.

# Name:

In [1]:
## RUN THIS CELL TO GET THE RIGHT FORMATTING AND TO LOAD NumPy
import requests
import numpy as np
from IPython.core.display import HTML
styles = requests.get("https://raw.githubusercontent.com/Harvard-IACS/2018-CS109A/master/content/styles/cs109.css").text
HTML(styles)

# Tasks

- Review pages 113-122 in the Course Notes.
- Submit completed Jupyter notebook on Moodle.

# Exercises

The file `tourism.csv` contains data collected from the Las Vegas tourism bureau--this should be in the same directory as this notebook.<br><br>

<div class="exercise"><b>Exercise 1</b></div> 

In the code cell below, write Python code to do the following.<br>

- Import the Pandas library
- Read in the file `tourism.csv` and assign it to the data frame `lv_df`.<br>
- Print out the first five rows using the `head()` function
- Take a couple of minutes to familiarize yourself with the data set

In [3]:
import pandas as pd
lv_df = pd.read_csv('tourism.csv')
lv_df.head()

date  visitors  convention  roomsavail  occupancy  avgrate  revperroom  \
0  Jan-2017   3508800      736000      148693      0.857   149.13   127.80441   
1  Feb-2017   3150500      620200      147866      0.856   116.73    99.92088   
2  Mar-2017   3783900      757400      147717      0.940   148.55   139.63700   
3  Apr-2017   3545100      505800      147566      0.907   123.52   112.03264   
4  May-2017   3602100      514600      147656      0.891   124.51   110.93841   

   roomnights  planepassengers  avgtraffic  gamingrevenue  
0     3951300          3676329       99390    926124000.0  
1     3545900          3432220      102967    825618000.0  
2     4303200          4206358      117771    857317000.0  
3     4016100          4062190      121743    763254000.0  
4     4078400          4208019      119812    860700000.0

<div class="exercise"><b>Exercise 2</b></div>

Assign `max_visitors` the maximum number of visitors in the data set--please display the result.

The output from my solution is:<br>
<code>
The maximum number of visitors was 3788800</code>

In [13]:
max_visitors = lv_df['visitors'].max()
print(f'The maximum number of visitors was {max_visitors}')

The maximum number of visitors was 3788800


<div class="exercise"><b>Exercise 3</b></div>

Write Python code in the cell below to do the following.

- Sort the data by the number of visitors (in descending order) and assign the new data frame to `lv_sorted`.
- Print the first **10** rows of `lv_sorted`, but only for the `date` and `visitors` columns.

The output from my solution is:<br>
<code>
        date  visitors
6   Jul-2017   3788800
2   Mar-2017   3783900
14  Mar-2018   3749800
26  Mar-2019   3697100
28  May-2019   3691100
30  Jul-2019   3685900
21  Oct-2018   3680600
18  Jul-2018   3659600
16  May-2018   3630400
29  Jun-2019   3607400
</code>

In [23]:
lv_sorted = lv_df.sort_values('visitors', ascending = False)
lv_sorted[['date', 'visitors']].head(10)

date  visitors
6   Jul-2017   3788800
2   Mar-2017   3783900
14  Mar-2018   3749800
26  Mar-2019   3697100
28  May-2019   3691100
30  Jul-2019   3685900
21  Oct-2018   3680600
18  Jul-2018   3659600
16  May-2018   3630400
29  Jun-2019   3607400

<div class="exercise"><b>Exercise 4</b></div>

In the code cell below, define a lambda function called `get_month()` that takes a string of the form in the `date` column of our dataframe, i.e., 'month-year', and returns just the <u>month name</u>.

For example, `get_year('Aug-2017')` would return `Aug`.

The output from my solution below is:<br>
<code>
'Jul'</code>

In [33]:
test = 'Jul-2017'
## Define your lambda function here
get_month = lambda x: x.split('-')[0]
get_month('Jul-2018')

'Jul'

<div class="exercise"><b>Exercise 5</b></div>

In the code cell below, define a lambda function called `get_year()` that takes a string of the form in the `date` column of our dataframe, i.e., 'month-year', and returns just the <u>year</u> *as an integer*.

For example, `get_year('Aug-2017')` would return `2017` as an integer.

The output from my solution below is:<br>
<code>
2018</code>

In [43]:
## Define your lambda function here
get_year = lambda x: int(x.split('-')[1])
get_year('Jul-2018')

2018

<div class="exercise"><b>Exercise 6</b></div>

Write Python code in the cell below to do the following.

- Using your lambda function `get_month()` and the `apply()` function, add a new column to `lv_sorted` called `month` that contains the month for each row.
- Using your lambda function `get_year()` and the `apply()` function, add a new column to `lv_sorted`called `year` that contains the year for each row.
- Print the first **10** rows of `lv_sorted`, but only for the `month`, `year`, and `visitors` columns. 

The output from my solution is:<br>
<code>
   month  year  visitors
6    Jul  2017   3788800
2    Mar  2017   3783900
14   Mar  2018   3749800
26   Mar  2019   3697100
28   May  2019   3691100
30   Jul  2019   3685900
21   Oct  2018   3680600
18   Jul  2018   3659600
16   May  2018   3630400
29   Jun  2019   3607400</code>

In [47]:
get_month = lambda x: x.split('-')[0]
get_year = lambda x: int(x.split('-')[1])
lv_sorted['month'] = lv_sorted['date'].apply(get_month)
lv_sorted['year'] = lv_sorted['date'].apply(get_year)
lv_sorted[['month', 'year', 'visitors']].head(10)

month  year  visitors
6    Jul  2017   3788800
2    Mar  2017   3783900
14   Mar  2018   3749800
26   Mar  2019   3697100
28   May  2019   3691100
30   Jul  2019   3685900
21   Oct  2018   3680600
18   Jul  2018   3659600
16   May  2018   3630400
29   Jun  2019   3607400

<div class="exercise"><b>Exercise 7</b></div>

Project columns `year`, `month`, `visitors`, `convention`, and `roomsavail` of `lv_sorted` and assign to `lv_df2`. Finish by printing the last 7 lines of the result. 

The output from my solution is:<br>
<code>
    year month  visitors  convention  roomsavail
12  2018   Jan   3393900      612900      146715
10  2017   Nov   3300500      530600      146222
23  2018   Dec   3267600      221800      147238
11  2017   Dec   3194800      289100      146993
25  2019   Feb   3189900      748600      148168
1   2017   Feb   3150500      620200      147866
13  2018   Feb   3130400      690000      146672</code>

In [49]:
lv_df2 = lv_sorted[['year', 'month', 'visitors', 'convention', 'roomsavail']]
lv_df2.tail(7)

year month  visitors  convention  roomsavail
12  2018   Jan   3393900      612900      146715
10  2017   Nov   3300500      530600      146222
23  2018   Dec   3267600      221800      147238
11  2017   Dec   3194800      289100      146993
25  2019   Feb   3189900      748600      148168
1   2017   Feb   3150500      620200      147866
13  2018   Feb   3130400      690000      146672

<div class="exercise"><b>Exercise 8</b></div>

Obtain a **`Series`** with the average number of visitors, convention, and roomsavail from `lv_df2`, assigning to `avg_vals`.  Print the result.

The output from my solution is:<br>
<code>
visitors      3.521177e+06
convention    5.546871e+05
roomsavail    1.476232e+05
dtype: float64</code>

In [51]:
avg_vals = lv_df2[['visitors', 'convention', 'roomsavail']].agg('mean')
print(avg_vals)

visitors      3.521177e+06
convention    5.546871e+05
roomsavail    1.476232e+05
dtype: float64


<div class="exercise"><b>Exercise 9</b></div>

Compute the average number of visitors, convention, and roomsavail **by year** of `lv_df2` in a `DataFrame` named `avg_vals_year`.  Print the result.

The output from my solution is:<br>
<code>
          visitors     convention     roomsavail
year                                            
2017  3.517842e+06  553850.000000  147267.500000
2018  3.509742e+06  541816.666667  147418.666667
2019  3.546500e+06  578185.714286  148583.571429</code>

In [57]:
avg_vals_year = lv_df2
print(avg_vals_year)

SyntaxError: unterminated string literal (detected at line 1) (1787805248.py, line 1)

<div class="exercise"><b>Exercise 10</b></div>

Compute the average number of visitors, convention, and roomsavail **by month** of `lv_df2` in a `DataFrame` named `avg_vals_month`.  Print the result.

The output from my solution is:<br>
<code>
           visitors     convention     roomsavail
month                                            
Apr    3.545033e+06  533033.333333  147824.666667
Aug    3.559050e+06  597450.000000  147045.500000
Dec    3.231200e+06  255450.000000  147115.500000
Feb    3.156933e+06  686266.666667  147568.666667
Jan    3.438267e+06  677433.333333  147864.666667
Jul    3.711433e+06  471266.666667  147796.333333
Jun    3.592833e+06  500300.000000  148174.666667
Mar    3.743600e+06  654000.000000  147728.666667
May    3.641200e+06  508166.666667  148176.000000
Nov    3.389500e+06  571950.000000  146883.500000
Oct    3.642450e+06  676400.000000  147090.000000
Sep    3.512100e+06  450700.000000  147324.500000</code>